overlord/servicestate/quota_control.go

```diff
@@ -68,8 +102,136 @@ func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[s
+			//    current limit.
+			// 3. If the slice was modified to be given _less_ resources (i.e. a
+			//    lower memory limit), then we need to stop the services before
+			//    issuing the daemon-reload to systemd, then do the
+			//    daemon-reload which will succeed in modifying the cgroup, then
+			//    start the services we stopped back up again. This is because
+			//    otherwise if the services are currently running and using more
+			//    resources than they would be allowed after the modification is
+			//    applied by systemd to the cgroup, the kernel responds with
+			//    EBUSY, and it isn't clear if the modification is then properly
+			//    in place or not.
+			//
+			// We will already have called daemon-reload at the end of
+			// EnsureSnapServices directly, so handling case 3 is difficult, and
+			// for now we disallow making this sort of change to a quota group,
+			// that logic is handled at a higher level than this function.
+			// Thus the only decision we really have to make is if the slice was
+			// newly written or not, and if it was save it for later
+			if old == "" {
+				grpsToStart = append(grpsToStart, grp)
+			}
+
+		case "service":
+			// in this case, the only way that a service could have been changed
+			// was if it was moved into or out of a slice, in both cases we need
+			// to restart the service
+			sn := app.Snap
+			appsToRestartBySnap[sn] = append(appsToRestartBySnap[sn], app)
+
+			// TODO: what about sockets and timers? activation units just start
+			// the full unit, so as long as the full unit is restarted we should
+			// be okay?
+		}
+	}
+	if err := wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, collectModifiedUnits, progress.Null); err != nil {
+		return err
+	}
+
+	// now start the slices
+
+	// TODO: should this logic move to wrappers in wrappers.RestartGroups() ?
+	systemSysd := systemd.New(systemd.SystemMode, progress.Null)
+
+	for _, grp := range grpsToStart {
+		// TODO: what should these timeouts for stopping/restart slices be?
+		if !ensureOpts.Preseeding {
+			if err := systemSysd.Start(grp.SliceFileName()); err != nil {
+				return err
+			}
+		}
+	}
+
+	// after starting all the grps that we modified from EnsureSnapServices,
+	// we need to handle the case where a quota was removed, this will only
+	// happen one at a time and can be identified by the grp provided to us not
+	// existing in the state
+	if _, ok := allGrps[grp.Name]; !ok {
+		// stop the quota group, then remove it
+		if !ensureOpts.Preseeding {
+			if err := systemSysd.Stop(grp.SliceFileName(), 5*time.Second); err != nil {
+				logger.Noticef("unable to stop systemd slice while removing group %q: %v", grp.Name, err)
+			}
+		}
+
+		// TODO: this results in a second systemctl daemon-reload which is
+		// undesirable, we should figure out how to do this operation with a
+		// single daemon-reload
+		if err := wrappers.RemoveQuotaGroup(grp, progress.Null); err != nil {
```

Comment:
we wouldn't remove a group at preseed time anyway

Reply:
yeah I made RemoveQuota just fail if we are preseeding anyways since it doesn't make sense to support that at all right now